### ISSUE TITLE
German translation almost finished: need clarification for a few things

### ISSUE THREAD

**Issue by tiguchi** [sqitchers/sqitch#449]
Hey, I just finished the German translation. Before I turn it into a PR I have a couple of questions for clarification:

# 1. "Please answer with y or n"

Is it possible to localize the answer options? In German it would be "j" instead of "y". 

# 2. "Out of order"

What's the context for that phrase? Can I see a full example? Is that a displaced item in an enumeration or does it refer to a failing service such as a DB connection that is currently down?

# 3. "No label passed to the _ format"

What's the context here? Can I see a full example? Does the underscore represent a placeholder? Is it substituted with something?

# Testing

How can I test-drive the German version of sqitch? I'm no Perl expert. How can I override the system locale so it gives me the German version?

**Comment by theory** [comment]
> Is it possible to localize the answer options? In German it would be "j" instead of "y".

We could, yes. It's just in the [`ask_y_n` method](https://github.com/sqitchers/sqitch/blob/1e6705bf6ffe379dd314429220f3c9842a5ec9c1/lib/App/Sqitch.pm#L424), which would need to be updated to fetch the proper characters to prompt for and search for. Do you want to include that change in your PR?

> What's the context for that phrase? Can I see a full example? Is that a displaced item in an enumeration or does it refer to a failing service such as a DB connection that is currently down?

It comes from [this line in App::Sqitch::Engine](https://github.com/sqitchers/sqitch/blob/1e6705bf6ffe379dd314429220f3c9842a5ec9c1/lib/App/Sqitch/Engine.pm#L469), which I think is used by `sqitch verify` when it finds changes are not listed in the proper order. Not sure why that happens anymore, though (why didn't I write more comments??).

> What's the context here? Can I see a full example? Does the underscore represent a placeholder? Is it substituted with something?

Yes, it comes from list of formats used by the [log command[(https://github.com/sqitchers/sqitch/blob/develop/lib/sqitch-log.pod#Formats) (and plan, IIRC).  It's the `%{$thing}_` placeholder, which is a localized label for something, e.g. `%{change}_ %H` would replace `%{change}_` with the localized label for "change", while `%H` would be the change ID. In the case of this error, it would trigger if someone used `%_` without a label inside braces, as in [this test](https://github.com/sqitchers/sqitch/blob/1e6705bf6ffe379dd314429220f3c9842a5ec9c1/t/log.t#L457).

> How can I test-drive the German version of sqitch? I'm no Perl expert. How can I override the system locale so it gives me the German version?

There are no tests for the localization. I would just set the `$LANG` environment variable and make sure it works. Er, unfortunately, that requires compiling the .po files, so don't worry about it, we'll make sure it works before merging the changes.

Thanks!

**Comment by tiguchi** [comment]
Thanks for the clarification!

> We could, yes. It's just in the [`ask_y_n` method](https://github.com/sqitchers/sqitch/blob/1e6705bf6ffe379dd314429220f3c9842a5ec9c1/lib/App/Sqitch.pm#L424), which would need to be updated to fetch the proper characters to prompt for and search for. Do you want to include that change in your PR?

Sure, but I'll leave that Perl function alone, I don't want to mess up stuff 😄. 

So how de we go about it? Would we add two extra translation file entries that are pulled in as answer options? Something like:

```
% Answer option for "y"es
msgid "y"
msgstr "j"

% Answer option for "n"o
msgid "n"
msgstr "n"
```

And the 3 regexes in that function would be constructed based on these translations?

**Comment by theory** [comment]
I'm working on a new method, `confirm()`, that localizes confirmation prompts. The new PO entries look like this:

```
#: lib/App/Sqitch.pm:428
msgctxt "Confirm prompt answer yes"
msgid "Yes"
msgstr ""

#: lib/App/Sqitch.pm:429
msgctxt "Confirm prompt answer no"
msgid "No"
msgstr ""
```

The `msgctxt` entries provide context, so you know the purpose of these otherwise super short messages. Does that work for you?

BTW, I always assumed the German negative response would be "nein". But that's one of, like, four words of German I know, so I'm not surprised the context is completely wrong.

**Comment by theory** [comment]
Duh, "j" is for "Ja", not the negation. Please ignore my ignorance.

**Comment by tiguchi** [comment]
Thanks for the quick update! Will the new prompt function also accept the first letter of each answer option, or will users have to spell out their response?

I guess spelling out the answer might help prevent accidental data loss when the user runs a revert, so I think it's good change. Other users might disagree though?

**Comment by tiguchi** [comment]
Ah nevermind... I just read your commit message and you already explain how it works:

> The user may input a substring of either answer, such as "y" for "Yes" or "j" for "Ja". 
> Translators therefore must be sure that the first char of each answer is distinct, at least for now.

**Comment by theory** [comment]
Should this issue be closed now that you've submitted #451, @tiguchi?

**Comment by tiguchi** [comment]
Yes, closing this issue now :-)
